libmamba/src/core/subdirdata.cpp

```diff
@@ -115,182 +166,226 @@ namespace mamba
+                fmt::format("File: {}: Could not load cache state: {}", state_file, e.what()),
+                mamba_error_code::cache_not_loaded
+            );
+        }
+
+        if (!m.check_valid_metadata(repodata_file))
+        {
+            LOG_WARNING << "Cache file " << repodata_file << " was modified by another program";
             return make_unexpected(
-                fmt::format("Could not load cache state: {}", e.what()),
+                fmt::format("File: {}: Cache file mtime mismatch", state_file),
                 mamba_error_code::cache_not_loaded
             );
         }
         return m;
     }
 
+    auto MSubdirMetadata::from_repodata_file(const fs::u8path& repodata_file)
+        -> expected_subdir_metadata
+    {
+        std::ifstream in_file = open_ifstream(repodata_file);
+        std::string json = extract_subjson(in_file);
 
-    namespace detail
+        try
+        {
+            nlohmann::json result = nlohmann::json::parse(json);
+            MSubdirMetadata m;
+            m.m_http.url = result.value("_url", "");
+            m.m_http.etag = result.value("_etag", "");
+            m.m_http.last_modified = result.value("_mod", "");
+            m.m_http.cache_control = result.value("_cache_control", "");
+            return m;
+        }
+        catch (std::exception& e)
+        {
+            LOG_WARNING << "Could not parse mod/etag header";
+            return make_unexpected(
+                fmt::format("File: {}: Could not parse mod/etag header ({})", repodata_file, e.what()),
+                mamba_error_code::cache_not_loaded
+            );
+        }
+    }
+
+    bool MSubdirMetadata::checked_at::has_expired() const
     {
-        tl::expected<subdir_metadata, mamba_error> read_metadata(const fs::u8path& file)
+        // difference in seconds, check every 14 days
+        return std::difftime(std::time(nullptr), last_checked) > 60 * 60 * 24 * 14;
+    }
+
+    /***************
+     * MSubdirData *
+     ***************/
+
+    namespace
+    {
+        using file_duration = fs::file_time_type::duration;
+        using file_time_point = fs::file_time_type::clock::time_point;
+
+        template <typename T>
+        std::vector<T> without_duplicates(std::vector<T>&& values)
         {
-            auto state_file = file;
-            state_file.replace_extension(".state.json");
-            std::error_code ec;
-            if (fs::exists(state_file, ec))
+            const auto end_it = std::unique(values.begin(), values.end());
+            values.erase(end_it, values.end());
+            return values;
+        }
+
+        file_duration get_cache_age(const fs::u8path& cache_file, const file_time_point& ref)
+        {
+            try
+            {
+                fs::file_time_type last_write = fs::last_write_time(cache_file);
+                file_duration tdiff = ref - last_write;
+                return tdiff;
+            }
+            catch (...)
```

Comment:
Or even better the proper exception related to the error?
```suggestion
            catch (std::exception const&)
```